farming/logics/traits/master_chef/farming.rs

```diff
@@ -0,0 +1,482 @@
+        Ok(())
+    }
+
+    #[ink(message)]
+    fn pending_arsw(&mut self, pool_id: u32, user: AccountId) -> Result<Balance, FarmingError> {
+        let pool = self
+            .get_pool_infos(pool_id)
+            .ok_or(FarmingError::PoolNotFound4)?;
+        let user_info = self
+            .get_user_info(pool_id, user)
+            .ok_or(FarmingError::UserNotFound)?;
+        let mut acc_arsw_per_share = pool.acc_arsw_per_share;
+
+        let lp_supply = self._get_lp_supply(pool_id)?;
+        let current_block = Self::env().block_number();
+
+        if current_block > pool.last_reward_block && lp_supply != 0 {
+            let additional_acc_arsw_per_share =
+                self._calculate_additional_acc_arsw_per_share(&pool, current_block, lp_supply)?;
+            acc_arsw_per_share = acc_arsw_per_share
+                .checked_add(additional_acc_arsw_per_share)
+                .ok_or(FarmingError::AddOverflow8)?;
+        }
+
+        let pending = <u128 as TryInto<i128>>::try_into(
+            user_info
+                .amount
+                .checked_mul(acc_arsw_per_share)
+                .ok_or(FarmingError::MulOverflow9)?
+                / ACC_ARSW_PRECISION,
+        )
+        .map_err(|_| FarmingError::CastToi128Error)?
+        .checked_sub(user_info.reward_debt)
+        .ok_or(FarmingError::SubUnderflow6)?;
+        Ok(<i128 as TryInto<u128>>::try_into(pending)
+            .map_err(|_| FarmingError::CastTou128Error1)?)
+    }
+
+    #[ink(message)]
+    fn deposit(
+        &mut self,
+        pool_id: u32,
+        amount: Balance,
+        to: AccountId,
+    ) -> Result<(), FarmingError> {
+        let pool = self
+            .get_pool_infos(pool_id)
+            .ok_or(FarmingError::PoolNotFound4)?;
+        self._update_pool(pool_id)?;
+        let user = self.get_user_info(pool_id, to).unwrap_or_default();
+        let user_amount = user
+            .amount
+            .checked_add(amount)
+            .ok_or(FarmingError::AddOverflow10)?;
+        let user_reward_debt = user
+            .reward_debt
+            .checked_add(
+                <u128 as TryInto<i128>>::try_into(
+                    amount
+                        .checked_mul(pool.acc_arsw_per_share)
+                        .ok_or(FarmingError::MulOverflow10)?
+                        / ACC_ARSW_PRECISION,
+                )
+                .map_err(|_| FarmingError::CastToi128Error2)?,
+            )
+            .ok_or(FarmingError::AddOverflow11)?;
+
+        self.data::<Data>().user_info.insert(
+            &(pool_id, to),
+            &UserInfo {
+                amount: user_amount,
+                reward_debt: user_reward_debt,
+            },
+        );
+
+        if let Some(rewarder_address) = self.get_rewarder(pool_id) {
+            if rewarder_address != ZERO_ADDRESS.into() {
+                RewarderRef::on_arsw_reward(&rewarder_address, to, to, user_amount)?;
```

Comment:
pid is missing as an argument of on_arsw_reward

Reply:
Based on mock:
```
    function onARSWReward(
        uint256,
        address ,
        address to,
        uint256 arswAmount,
        uint256
    ) external override onlyMasterChef {
        uint256 pendingReward = arswAmount.mul(rewardMultiplier) /
            REWARD_TOKEN_DIVISOR;
        uint256 rewardBal = rewardToken.balanceOf(address(this));
        if (pendingReward > rewardBal) {
            rewardToken.safeTransfer(to, rewardBal);
        } else {
            rewardToken.safeTransfer(to, pendingReward);
        }
    }
```
This:  r`ewarder.onARSWReward(pid, to, to, 0, user.amount);`
Will have no effect ( as 0 is `arswAmount` )

Still updated